### PR TITLE
Add config to disable logging row data from processing log + config docs

### PR DESCRIPTION
**Pull request by rodesai**
### Description 

    This patch adds a config to disable logging actual row data from
    the processing log. The config will be defaulted to true. The
    reasoning for this is that its dangerous from a security pov to
    log row data, so users should be forced to do configure this
    behaviour explicitly.

    From an implementation POV, this is used by making the processing
    log config a global, and accessing it from the various message
    factories. We should also consolidate all the processing-log related
    code to its own module, but thats deferred to a follow-on to keep
    this PR small.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/processing-log.rst [hunk 1]
```diff
@@ -65,11 +65,15 @@ Note that the server must be restarted for the config to take effect.
 Note On Security
 ================
 
-Entries in the record-processing log may include data from the rows processed by KSQL.
-Be careful to ensure that the log is configured to write to a destination
-where it is safe to write the data being processed. It's also important to set
-``log4j.additivity.processing=false`` as shown in the previous example, to ensure that processing log
-events are not forwarded to appenders configured for the other KSQL loggers.
+By default, the record-processing log won't log any actual row data. To help
```
Comment (JimGalasyn) [inline docs/developer-guide/processing-log.rst]:
```suggestion
By default, the record-processing log doesn't log any actual row data. To help
```

File: docs/developer-guide/processing-log.rst [hunk 2]
```diff
@@ -65,11 +65,15 @@ Note that the server must be restarted for the config to take effect.
...
+you debug, you can enable including row data in log messages by setting the
+KSQL property ``ksql.processing.log.include.rows`` to ``true``.
+
+If you do this, you should be careful to ensure that the log is configured to
```
Comment (JimGalasyn) [inline docs/developer-guide/processing-log.rst]:
```suggestion
If you do this, ensure that the log is configured to
```

File: docs/installation/server-config/config-reference.rst [hunk 3]
```diff
@@ -211,6 +211,82 @@ bind to the default interface. For example:
...
+If automatic processing log topic creation is enabled, KSQL creates the topic with  number of replicas set
+to the value of this property. By default, this property has the value ``1``.
+
+.. _ksql-processing-log-stream-auto-create:
+
+--------------------------------------
+ksql.processing.log.stream.auto.create
+--------------------------------------
+
+Toggles automatic processing log stream creation. If set to true, and running interactive mode on a new cluster,
```
Comment (JimGalasyn) [inline docs/installation/server-config/config-reference.rst]:
```suggestion
Toggles automatic processing log stream creation. If set to true, and KSQL is running in interactive mode on a new cluster,
```

File: docs/installation/server-config/config-reference.rst [hunk 4]
```diff
@@ -211,6 +211,82 @@ bind to the default interface. For example:
...
+then KSQL will automatically create a processing log stream when it starts up. The name for the stream is the
```
Comment (JimGalasyn) [inline docs/installation/server-config/config-reference.rst]:
```suggestion
KSQL automatically creates a processing log stream when it starts up. The name for the stream is the
```

File: docs/installation/server-config/config-reference.rst [hunk 5]
```diff
@@ -211,6 +211,82 @@ bind to the default interface. For example:
...
+value of the :ref:`ksql-processing-log-stream-name` property. The stream will be created over the topic set in
```
Comment (JimGalasyn) [inline docs/installation/server-config/config-reference.rst]:
```suggestion
value of the :ref:`ksql-processing-log-stream-name` property. The stream is created over the topic set in
```

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java [hunk 6]
```diff
@@ -24,6 +26,23 @@
 public class ProcessingLogConfig extends AbstractConfig {
   private static final String PROPERTY_PREFIX = "processing.log.";
 
+  private static ProcessingLogConfig INSTANCE = null;
+
+  public static void configure(final Map<?, ?> properties) {
+    Objects.requireNonNull(properties);
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java]: nit: unnecessary null check.  It's checked down the stack anyway.

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java [hunk 7]
```diff
@@ -24,6 +26,23 @@
 public class ProcessingLogConfig extends AbstractConfig {
   private static final String PROPERTY_PREFIX = "processing.log.";
 
+  private static ProcessingLogConfig INSTANCE = null;
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java]: statics.... are nasty.  We should do everything we can to avoid having to rely on statics. Is there _really_ no alternative?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutor.java [hunk 8]
```diff
@@ -27,7 +27,7 @@
 import io.confluent.ksql.parser.tree.SetProperty;
 import io.confluent.ksql.parser.tree.Statement;
 import io.confluent.ksql.parser.tree.UnsetProperty;
-import io.confluent.ksql.rest.util.ProcessingLogConfig;
+import io.confluent.ksql.processing.log.ProcessingLogConfig;
```
Comment (big-andy-coates) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutor.java]: nit: can we switch this around to `logging.processing`, then we can have other logging stuff under `logging.blah` etc.

File: ksql-serde/src/main/java/io/confluent/ksql/serde/util/SerdeProcessingLogMessageFactory.java [hunk 9]
```diff
@@ -30,17 +31,30 @@ private SerdeProcessingLogMessageFactory() {
   public static Supplier<SchemaAndValue> deserializationErrorMsg(
       final Throwable exception,
       final Optional<byte[]> record) {
+    return deserializationErrorMsg(
+        exception,
+        record,
+        ProcessingLogConfig.getInstance()
```
Comment (big-andy-coates) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/util/SerdeProcessingLogMessageFactory.java]: This type is only called from the deserialisers - so we don't need a static `getInstance` for this. The deserializer can grab the processing config out of the `configure` method, right?

File: docs/installation/server-config/config-reference.rst [hunk 10]
```diff
@@ -211,6 +211,82 @@ bind to the default interface. For example:
...
 
+KSQL Processing Log Settings
+----------------------------
+
+These configurations control the behavior of the :ref:`KSQL processing log <ksql_processing_log>`.
+
+.. _ksql-processing-log-topic-auto-create:
+
+-------------------------------------
+ksql.processing.log.topic.auto.create
```
Comment (big-andy-coates) [inline docs/installation/server-config/config-reference.rst]: we should be generating such docs _from_ the config... :(

(Though don't get me wrong - you adding the docs is great!)

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java (hunk 7 above)
Comment (rodesai) [reply ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java]: Yeah, I don't love it either. I felt like in this case it might not be so bad since its usage is pretty contained, and its immutable (mostly, a test could still call configureFinal - couldn't think of a good way of preventing that), so you won't have the usual problems with unpredictable/untraceable behavior. The only things that need to modify the values are the processing log related tests, and we can pass a config in directly in those places. I was planning to enforce this more strictly (in a follow-up to keep the diff small) by moving everything that calls getInstance into another package (e.g. ksql.processing.log), and making getInstance package private.

The alternative would be to pass it all the way through to the  physical plan building stuff in the structured package. So we'd either have to:
    - Add these properties to KsqlConfig and pull them out of there, which I didn't want to do since KsqlConfig is becoming kind of a dumping ground for configs. We should refactor it to pull out the bits that track different AbstractConfig implementations into a parent manager class that can be configured to handle each implementation independently and be told which configs are compatibility-breaking or not.
     - Add it as a parameter everywhere.
     - Pass around a ConfigContext with the ksql config and this config.

I'm not super keen on any of the above, so I'm all ears to suggestions.

File: docs/installation/server-config/config-reference.rst (hunk 10 above)
Comment (rodesai) [reply docs/installation/server-config/config-reference.rst]: I looked at doing this at first and then punted on it. ConfigDef is set up to do this, but we need to do a few things:
- Integrate it into the sphinx build stuff, which is a fair bit of work - I think we'll need to add a build target that can generate the docs, then update the docs repo to use it for ksql and copy the artifact over to the ksql docs before running sphinx-build
- The format of the output is different from the rest of the config docs, so we'd want to do this for all the config docs. But we'd want to be sure to configure internal configs correctly in our code.

I think it makes sense to at least have docs first.

cc @JimGalasyn

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutor.java (hunk 8 above)
Comment (rodesai) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutor.java]: Sounds good. I'll do it in a follow up. I want to move some of the files/code around anyway.

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java (hunk 7 above)
Comment (big-andy-coates) [reply ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java]: KsqlConfig is a dumping ground - you're right.  But I don't think we're making the situation any better by ignoring it and using statics to get around the problem.  That's just added technical debt in my book.  

Maybe this is the right time to take a good look at our use of KsqlConfig?  Many methods already take both a `KsqlConfig` param and a `overrriddenProperties` param, and now we've got a processing log config.  

- `KsqlConfig` looks to be pretty much constant, i.e. the application creates it once at startup.  So why do we pass it around with each statement execution?  Could this not be injected at KsqlEngine creation?
- `overriddenProperties` can then be the bit that gets passed around per-execution.

That leaves the processing log properties.  Personally, I think we _should_ have a single top-level config class, `KsqlConfig` is probably a good name, but this should internally hold different instance of different config classes for the different areas, e.g. Ksql engine, streams and log processing.   

This top level class should also clearly define what the `overriddenProperties` should and should not be applied to. (i.e. SSL config and log processing properties should not be overridable).

In short, I feel that the log processing properties _should_ be part of some top level config.  Maybe now is the time to do a PR to refactor our current config usage to better support the addition of the processing log config...

What do you think? I'd even be willing to help if it meant we avoid statics! ;)

File: docs/installation/server-config/config-reference.rst (hunk 10 above)
Comment (big-andy-coates) [reply docs/installation/server-config/config-reference.rst]: > I think it makes sense to at least have docs first.

Absolutely. Just raising this as its not something that's getting any attention. It's always a 'would it be nice if...'. 

Added https://github.com/confluentinc/ksql/issues/2423

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java (hunk 7 above)
Comment (rodesai) [reply ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java]: > * `KsqlConfig` looks to be pretty much constant, i.e. the application creates it once at startup.  So why do we pass it around with each statement execution?  Could this not be injected at KsqlEngine creation?

Sure, but we still need to override with the persisted configs (for compatibility), then the user's  overrides before using it. So we have to pass around _something_, which IMO should be KsqlConfig. We can add a method for getting the overrides from it so we can return that info from that rest api (which is all they're used for after the call to `cloneWithPropertyOverwrites`. That way we don't have to pass the overrides around separately.
 
> Personally, I think we _should_ have a single top-level config class, `KsqlConfig` is probably a good name, but this should internally hold different instance of different config classes for the different areas

Agreed. This is pretty much that parent manager class I mentioned.

I don't want to block this change on the whole config refactor, since it should get into the next release. As a first step, I'll wrap the ksql config, overridden properties, and processing log config into another class, and we can pass that around. Then we can evolve that class into the thing that manages all the subsystem configs.

Comment (rodesai) [reply ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java]: I opted to pass around a context object for the processing log. Currently you can use the context to get to the logger factory and the config. In a follow-up I'll move the msg factories into it, and drop the public interface to get the config. So the pattern can just be that you initialize the context with the config when the process starts up, and use it to do all your processing log stuff. This also gets rid of the global logger factory.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 11]
```diff
@@ -184,12 +186,14 @@ public SchemaKStream into(
...
+      final QueryContext.Stacker contextStacker,
+      final ProcessingLogContext processingLogContext) {
     final Selection selection = new Selection(
         selectExpressions,
-        ProcessingLoggerFactory.getLogger(
+        processingLogContext.getLoggerFactory().getLogger(
             QueryLoggerUtil.queryLoggerName(
-                contextStacker.push(Type.PROJECT.name()).getQueryContext())));
+                contextStacker.push(Type.PROJECT.name()).getQueryContext())),
+        processingLogContext);
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: Is it standard practice to pass around both the context for an object and the object itself (in this case, `processingLogContext` and the logger itself)? Since each logger is only used with a single context, it feels like there should be a way to combine the two at the time the logger is created so they wouldn't need to be passed around separately. Not sure whether this is preferable though.

File: ksql-serde/src/test/java/io/confluent/ksql/serde/delimited/KsqlDelimitedDeserializerTest.java [hunk 12]
```diff
@@ -53,7 +55,10 @@ public void before() {
         .field("itemid".toUpperCase(), org.apache.kafka.connect.data.Schema.OPTIONAL_STRING_SCHEMA)
         .field("orderunits".toUpperCase(), org.apache.kafka.connect.data.Schema.OPTIONAL_FLOAT64_SCHEMA)
         .build();
-    delimitedDeserializer = new KsqlDelimitedDeserializer(orderSchema, recordLogger);
+    delimitedDeserializer = new KsqlDelimitedDeserializer(
+        orderSchema,
+        recordLogger,
+        ProcessingLogContext.create());
```
Comment (vcrfxia) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/delimited/KsqlDelimitedDeserializerTest.java]:
```suggestion
        processingLogContext);
```

File: ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java [hunk 13]
```diff
@@ -172,7 +174,8 @@ public void init() {
         new KsqlConfig(Collections.emptyMap()),
         false,
         MockSchemaRegistryClient::new,
-        "test");
+        "test",
+        ProcessingLogContext.create());
```
Comment (vcrfxia) [inline ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java]:
```suggestion
        processingLogContext);
```
Also in two more places below (in this same file).

File: ksql-serde/src/test/java/io/confluent/ksql/serde/util/SerdeProcessingLogMessageFactoryTest.java [hunk 14]
```diff
@@ -25,33 +25,36 @@
...
 import io.confluent.ksql.processing.log.ProcessingLogMessageSchema.MessageType;
 import io.confluent.ksql.serde.util.SerdeProcessingLogMessageFactory;
 import java.util.Base64;
+import java.util.Collections;
 import java.util.Optional;
 import java.util.stream.IntStream;
 import org.apache.kafka.connect.data.Schema;
 import org.apache.kafka.connect.data.SchemaAndValue;
 import org.apache.kafka.connect.data.Struct;
+import org.junit.After;
```
Comment (vcrfxia) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/util/SerdeProcessingLogMessageFactoryTest.java]: nit: unused import. (Also a few more in this file, and in `EngineProcessingLogMessageFactoryTest.java` and `PhysicalPlanBuilderTest.java`

File: ksql-engine/src/test/java/io/confluent/ksql/structured/SqlPredicateTest.java [hunk 15]
```diff
@@ -154,10 +154,6 @@ public void shouldWriteProcessingLogOnError() throws IOException {
             "Error evaluating predicate (TEST1.COL0 > 100): "
                 + "Invalid field type. Value must be Long.")
     );
-    final String rowString =
-        errorStruct.getString(ProcessingLogMessageSchema.RECORD_PROCESSING_ERROR_FIELD_RECORD);
-    final List<Object> row = (List) MAPPER.readValue(rowString, List.class);
```
Comment (vcrfxia) [inline ksql-engine/src/test/java/io/confluent/ksql/structured/SqlPredicateTest.java]: super nit: test no longer needs `throw IOException` with this removal.

File: docs/installation/server-config/config-reference.rst [hunk 16]
```diff
@@ -243,6 +243,82 @@ You can access KSQL Server by using |c3|. For more information, see
...
+ksql.processing.log.topic.replication.factor
+--------------------------------------------
+
+If automatic processing log topic creation is enabled, KSQL creates the topic with  number of replicas set
+to the value of this property. By default, this property has the value ``1``.
+
+.. _ksql-processing-log-stream-auto-create:
+
+--------------------------------------
+ksql.processing.log.stream.auto.create
```
[the remark this answers is not in the document]
Comment (rodesai) [reply docs/installation/server-config/config-reference.rst]: The user could already have a topic around they want to use. I think right now we log an error. I'll add some documentation that points out that the topic needs to be created one way or another.

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogContext.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+   * @return A factory for creating structured loggers for logging processing log records.
+   */
+  StructuredLoggerFactory getLoggerFactory();
+
+  /**
+   * Creates a processing log context that uses the supplied config.
+   * @param config the processing log config
+   * @return: A processing log context that uses the supplied config.
+   */
+  static ProcessingLogContextImpl create(final ProcessingLogConfig config) {
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogContext.java]: I'm trying to move towards putting the processing logger stuff in its own sub-project in ksql. I want it to have well defined interface, and then hide all the implementation details behind the interface. This way, we can have a well-defined way stuff outside the sub-project interacts with the sub-project thats totally decoupled from its internal interactions.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java (hunk 11 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: I think standard practice is kind of orthogonal to whether it makes sense to do this here or not. I don't know if context is the right word for this object. It's an object that represents our interface to the processing logger module configured according to a given config. I don't see the issue with passing around both if the SelectValueMapper needs to use them both. I guess we could just pass the query context down instead of the logger and create that from SelectValueMapper. I'm kind of ambivalent about which is better.

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogContext.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    return new ProcessingLogContextImpl(config);
+  }
+
+  /**
+   * Creates a processing log context that uses the default processing log config.
+   * @return A processing log context that uses the default config
+   */
+  static ProcessingLogContextImpl create() {
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogContext.java]: Its used by some clients that use our serdes (e.g. datagen)

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 19]
```diff
@@ -761,27 +784,34 @@ private static Query addInto(
   private static final class SandboxedExecutionContext implements KsqlExecutionContext {
 
     private final EngineContext engineContext;
+    private final ProcessingLogContext processingLogContext;
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: It still needs the context to set up the query. The loggers themselves could/should have this behavior though.

[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: I'd need to fix some stuff in common to make this work though. Would rather not block this PR on it.

File: ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogContext.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogContext.java]: +1 for the return type. The idea here was to have a standard implementation, but not have users of the processing log module worry about the actual class.
